File: packages/sdk-typescript/src/flush.ts
/**
 * Flush policy + async dispatcher for the Relay TypeScript SDK (W4.2).
 *
 * Parity with the Python ``relay.flush`` module
 * (``packages/sdk-python/relay/flush.py``). The Relay SDK MUST be safe to
 * call in production: a slow or unreachable sidecar must never block the
 * caller (VAL-W3-018 / W4 parity) or crash the host application
 * (VAL-W4-018). This module owns:
 *
 *   * :class:`FlushPolicy` -- the user-facing object with two knobs:
 *       ``mode``     -- ``sync`` or ``async``.
 *       ``onError``  -- ``raise`` or ``drop_and_log``.
 *   * :class:`AsyncFlushDispatcher` -- a sequential background dispatcher
 *     that serialises outbound HTTP requests when ``mode='async'`` so the
 *     caller's ``run.close()`` returns immediately.
 *
 * Per VAL-W4-018 the default flush policy is
 * ``{mode: 'async', onError: 'drop_and_log'}``. Background flush failures
 * MUST log a structured error envelope to stderr and MUST NEVER raise
 * into the application's request path.
 *
 * ASCII-only per CLAUDE.md "ASCII-Safe Source".
 */

import { RelayConfigError } from "./errors.js";

export type FlushMode = "sync" | "async";
export type OnErrorMode = "raise" | "drop_and_log";

export interface FlushPolicyShape {
  readonly mode: FlushMode;
  readonly onError: OnErrorMode;
}

/**
 * SDK flush policy (VAL-W4-018).
 *
 * The default is ``{mode: 'async', onError: 'drop_and_log'}`` so a slow
 * or unreachable sidecar never blocks the caller and never crashes the
 * host application. Callers needing strict synchronous semantics
 * (e.g. tests) opt in via ``new FlushPolicy({mode: 'sync', onError:
 * 'raise'})``.
 */
export class FlushPolicy implements FlushPolicyShape {
  readonly mode: FlushMode;
  readonly onError: OnErrorMode;

  constructor(options: Partial<FlushPolicyShape> = {}) {
    const mode = options.mode ?? "async";
    const onError = options.onError ?? "drop_and_log";
    if (mode !== "sync" && mode !== "async") {
      throw new RelayConfigError(
        `flush_policy.mode must be 'sync' or 'async'; received ${JSON.stringify(mode)}`,
        { details: { field: "mode", received: mode } },
      );
    }
    if (onError !== "raise" && onError !== "drop_and_log") {
      throw new RelayConfigError(
        `flush_policy.on_error must be 'raise' or 'drop_and_log'; received ${JSON.stringify(onError)}`,
        { details: { field: "onError", received: onError } },
      );
    }
    this.mode = mode;
    this.onError = onError;
  }

  /** Coerce a user-supplied object / FlushPolicy / undefined into one. */
  static fromInput(value: unknown): FlushPolicy {
    if (value === undefined || value === null) return new FlushPolicy();
    if (value instanceof FlushPolicy) return value;
    if (typeof value !== "object" || Array.isArray(value)) {
      throw new RelayConfigError("flush_policy must be an object or FlushPolicy", {
        details: { received_type: typeof value },
      });
    }
    const obj = value as Record<string, unknown>;
    const allowed = new Set(["mode", "onError"]);
    const unknown = Object.keys(obj).filter((k) => !allowed.has(k));
    if (unknown.length > 0) {
      throw new RelayConfigError(
        `flush_policy has unknown key(s): ${JSON.stringify(unknown.sort())}`,
        { details: { unknown_keys: unknown.sort(), allowed_keys: [...allowed].sort() } },
      );
    }
    const partial: { mode?: FlushMode; onError?: OnErrorMode } = {};
    if (obj["mode"] !== undefined) partial.mode = obj["mode"] as FlushMode;
    if (obj["onError"] !== undefined) partial.onError = obj["onError"] as OnErrorMode;
    return new FlushPolicy(partial);
  }
}

export type AsyncWork = () => Promise<void>;

/**
 * Sequential background dispatcher for ``mode='async'`` flushes.
 *
 * Lazily-started: the worker promise chain is created on first
 * :meth:`submit`. A call to :meth:`submit` enqueues the work and returns
 * synchronously; the dispatcher's worker runs each submitted callable
 * in order via Promise chaining.
 *
 * A failure inside the callable is caught and converted according to the
 * configured :class:`FlushPolicy.onError`:
 *
 *   * ``raise``: the failure is recorded on :attr:`lastError` and an
 *     ERROR-level structured log line is written to stderr; subsequent
 *     awaits of :meth:`waitIdle` resolve normally so the dispatcher can
 *     drain.
 *   * ``drop_and_log``: a single WARN-level structured envelope is
 *     written to stderr and the failure is otherwise swallowed
 *     (VAL-W4-018).
 *
 * The dispatcher is stopped via :meth:`close` (idempotent).
 */
export class AsyncFlushDispatcher {
  readonly onError: OnErrorMode;
  private chain: Promise<void> = Promise.resolve();
  private closed = false;
  /** Last error captured under ``onError='raise'`` mode (test seam). */
  lastError: unknown = null;
  /** Test-observability: count of items the worker has processed. */
  processedCount = 0;
  /** Test-observability: count of dropped items (drop_and_log mode). */
  dropCount = 0;
  /** Test-observability: count of errors recorded under raise mode. */
  errorCount = 0;
  /** Optional injected stderr writer for tests. Defaults to ``process.stderr``. */
  private readonly stderrWrite: (line: string) => void;

  constructor(options: {
    onError?: OnErrorMode;
    stderrWrite?: (line: string) => void;
  } = {}) {
    this.onError = options.onError ?? "drop_and_log";
    this.stderrWrite =
      options.stderrWrite ??
      ((line: string) => {
        // node:stream's writable.write returns a boolean; ignore.
        process.stderr.write(line.endsWith("\n") ? line : `${line}\n`);
      });
  }

  /**
   * Enqueue ``work`` for the background worker.
   *
   * Returns immediately; the only synchronous cost is the chained
   * Promise creation (tens of microseconds). The caller is guaranteed
   * not to block on outbound HTTP I/O.
   *
   * If the dispatcher is already closed, ``submit`` runs the work
   * inline as a degraded-mode fallback (best-effort; the on_error
   * policy still governs failure propagation).
   */
  submit(work: AsyncWork): void {
    if (this.closed) {
      // Fire-and-forget: do not return the promise; the caller has
      // already exited the context.
      void this.runOne(work);
      return;
    }
    this.chain = this.chain.then(() => this.runOne(work));
  }

  /** Block until every submitted item has been processed. */
  async waitIdle(): Promise<void> {
    await this.chain;
  }

  /** Stop accepting new work. The chain is allowed to finish. */
  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    // Allow any pending chain to settle. The caller may choose to
    // ``await`` or fire-and-forget depending on the flush policy.
    await this.chain;
  }

  // -- internals ---------------------------------------------------------

  private async runOne(work: AsyncWork): Promise<void> {
    try {
      await work();
      this.processedCount += 1;
    } catch (err) {
      if (this.onError === "drop_and_log") {
        this.dropCount += 1;
        this.emitStructured("warning", "RELAY-SDK-FLUSH-DROP-AND-LOG", err);
        return;
      }
      this.errorCount += 1;
      this.lastError = err;
      this.emitStructured("error", "RELAY-SDK-FLUSH-ERROR", err);
    }
  }

  private emitStructured(level: "warning" | "error", code: string, err: unknown): void {
    const payload = {
      schema_version: "relay.error.v1",
      level,
      code,
      message: err instanceof Error ? err.message : String(err),
      error_class: err instanceof Error ? err.name : "UnknownError",
      details: {
        on_error: this.onError,
      },
    };
    try {
      this.stderrWrite(JSON.stringify(payload));
    } catch {
      // Stderr write must never throw into the host application.
    }
  }
}

File: packages/sdk-python/relay/flush.py
"""Flush policy + async dispatcher for the Relay Python SDK (W3.2).

The Relay SDK MUST be safe to call in production: a slow or unreachable
sidecar must never block the calling thread (VAL-W3-018) or crash the
host application (VAL-W3-019). This module owns:

  * :class:`FlushPolicy` -- the user-facing dataclass with two knobs:
      ``mode``     -- ``sync`` or ``async``.
      ``on_error`` -- ``raise`` or ``drop_and_log``.
  * :class:`AsyncFlushDispatcher` -- a single-threaded background worker
    that serializes outbound HTTP requests when ``mode='async'`` so the
    caller's ``trace.__exit__`` returns immediately.

Per VAL-W3-018 the dispatcher does NOT use any wall-clock threshold. The
test contract is a synchronization marker: ``__exit__`` returns BEFORE
the slow sidecar handler has even ``accept()``-ed the inbound socket.
This is achieved by enqueueing the call onto a ``queue.Queue`` and
returning -- the dispatcher's worker thread pulls the call and runs it
asynchronously.

Per VAL-W3-019 the dispatcher under ``on_error='drop_and_log'`` emits a
single ``WARN``-level log line on transport failure and continues; no
exception is propagated to the host application.

ASCII-only per CLAUDE.md "ASCII-Safe Source".
"""

from __future__ import annotations

import logging
import queue
import threading
from collections.abc import Callable
from dataclasses import dataclass
from typing import Any, Final, Literal

logger = logging.getLogger("relay.flush")

FlushMode = Literal["sync", "async"]
OnErrorMode = Literal["raise", "drop_and_log"]

_DEFAULT_FLUSH_QUEUE_DEPTH: Final[int] = 1024


@dataclass(frozen=True)
class FlushPolicy:
    """SDK flush policy (spec line 2016).

    Attributes:
        mode: ``sync`` -- ``trace.__exit__`` blocks on outbound HTTP I/O
            until the sidecar acknowledges. ``async`` -- the call is
            enqueued onto a background dispatcher thread and
            ``trace.__exit__`` returns immediately (VAL-W3-018).
        on_error: ``raise`` -- a transport failure propagates as a
            :class:`relay.errors.RelayError` subclass. ``drop_and_log``
            -- a transport failure emits one WARN log line and is
            otherwise swallowed (VAL-W3-019).
    """

    mode: FlushMode = "sync"
    on_error: OnErrorMode = "raise"

    @classmethod
    def from_mapping(cls, value: Any) -> FlushPolicy:
        """Coerce a user-supplied mapping or :class:`FlushPolicy` into one.

        Accepts:
          - ``None`` -> defaults ``(sync, raise)``.
          - a :class:`FlushPolicy` instance (returned as-is).
          - a dict with ``mode`` and/or ``on_error`` keys; unknown keys
            raise :class:`relay.errors.RelayConfigError`.
        """
        from .errors import RelayConfigError

        if value is None:
            return cls()
        if isinstance(value, FlushPolicy):
            return value
        if not isinstance(value, dict):
            raise RelayConfigError(
                "flush_policy must be a dict or FlushPolicy",
                details={"received_type": type(value).__name__},
            )
        allowed = {"mode", "on_error"}
        unknown = sorted(set(value) - allowed)
        if unknown:
            raise RelayConfigError(
                f"flush_policy has unknown key(s): {unknown!r}",
                details={"unknown_keys": unknown, "allowed_keys": sorted(allowed)},
            )
        mode = value.get("mode", "sync")
        on_error = value.get("on_error", "raise")
        if mode not in ("sync", "async"):
            raise RelayConfigError(
                f"flush_policy.mode must be 'sync' or 'async'; received {mode!r}",
                details={"field": "mode", "received": mode},
            )
        if on_error not in ("raise", "drop_and_log"):
            raise RelayConfigError(
                "flush_policy.on_error must be 'raise' or 'drop_and_log'; "
                f"received {on_error!r}",
                details={"field": "on_error", "received": on_error},
            )
        return cls(mode=mode, on_error=on_error)


class AsyncFlushDispatcher:
    """Single-threaded background dispatcher for ``mode='async'`` flushes.

    Lazily-started: the worker thread is created on first ``submit``.
    A call to :meth:`submit` enqueues the work and returns immediately
    (the only blocking is a microsecond-scale ``Queue.put``).

    The worker thread runs each submitted callable in order. A transport
    failure inside the callable is caught and converted according to the
    configured :class:`FlushPolicy.on_error`:

      * ``raise``: the exception is logged at ERROR level AND the
        :class:`AsyncFlushDispatcher` records it on
        :attr:`last_error` so callers can drain via :meth:`wait_idle`
        and inspect.
      * ``drop_and_log``: the exception is logged at WARN level and
        otherwise swallowed -- the host application is not perturbed
        (VAL-W3-019).

    The dispatcher is stopped explicitly via :meth:`close`, which signals
    the worker via a sentinel and joins it. The worker thread is a
    daemon thread so an un-closed dispatcher does not block interpreter
    shutdown.
    """

    _SENTINEL: Final[object] = object()

    def __init__(
        self,
        *,
        on_error: OnErrorMode = "raise",
        max_depth: int = _DEFAULT_FLUSH_QUEUE_DEPTH,
    ) -> None:
        self._on_error: OnErrorMode = on_error
        self._queue: queue.Queue[Any] = queue.Queue(maxsize=max_depth)
        self._worker: threading.Thread | None = None
        self._started_lock = threading.Lock()
        self._closed = False
        self.last_error: BaseException | None = None
        # Test-observability: counter of items the worker has processed.
        self.processed_count: int = 0
        # Test-observability: count of dropped (on_error='drop_and_log') items.
        self.drop_count: int = 0

    @property
    def on_error(self) -> OnErrorMode:
        return self._on_error

    def submit(self, work: Callable[[], Any]) -> None:
        """Enqueue ``work`` for the background worker.

        Returns immediately -- the only blocking is the ``Queue.put``
        which holds a lock for tens of microseconds. The caller is
        guaranteed not to block on outbound HTTP I/O (VAL-W3-018).

        If the dispatcher is already closed, ``submit`` runs the work
        inline. This is a degenerate degraded-mode path that should
        only occur during teardown.
        """
        if self._closed:
            # Cannot enqueue after close. Execute inline (best-effort)
            # so the call is not silently dropped at the SDK boundary;
            # ``on_error`` still governs whether failure propagates.
            self._run_one(work)
            return
        self._ensure_worker()
        self._queue.put(work)

    def wait_idle(self, *, timeout: float | None = None) -> bool:
        """Block until every submitted item has been processed.

        Returns True on idle; False if ``timeout`` elapsed first.
        Used by tests after they have released the slow handler.
        """
        if self._worker is None:
            return True
        # Queue.join honors task_done; we call task_done from the worker
        # AFTER each item completes.
        joined = threading.Event()

        def _waiter() -> None:
            self._queue.join()
            joined.set()

        t = threading.Thread(target=_waiter, daemon=True, name="relay-flush-wait")
        t.start()
        return joined.wait(timeout=timeout)

    def close(self, *, timeout: float = 30.0) -> None:
        """Drain the queue, stop the worker, and join it.

        Idempotent. After ``close`` returns the dispatcher is in a
        closed state and further ``submit`` calls run inline.
        """
        if self._closed:
            return
        self._closed = True
        worker = self._worker
        if worker is None:
            return
        # Push the sentinel; the worker loop exits on receiving it.
        self._queue.put(self._SENTINEL)
        worker.join(timeout=timeout)

    # -- internals -----------------------------------------------------------

    def _ensure_worker(self) -> None:
        if self._worker is not None:
            return
        with self._started_lock:
            if self._worker is not None:
                return
            self._worker = threading.Thread(
                target=self._worker_loop,
                name="relay-flush-dispatcher",
                daemon=True,
            )
            self._worker.start()

    def _worker_loop(self) -> None:
        while True:
            item = self._queue.get()
            try:
                if item is self._SENTINEL:
                    return
                self._run_one(item)
            finally:
                self._queue.task_done()

    def _run_one(self, work: Callable[[], Any]) -> None:
        try:
            work()
        except BaseException as exc:  # noqa: BLE001 - on_error governs disposition
            if self._on_error == "drop_and_log":
                # VAL-W3-019: WARN-level structured log line, NOT a stack
                # trace to stderr; the host application is not perturbed.
                self.drop_count += 1
                logger.warning(
                    "relay.flush.drop_and_log: dropped envelope; "
                    "error_type=%s error=%s",
                    type(exc).__name__,
                    str(exc),
                )
                return
            # 'raise' mode: record on last_error so wait_idle callers can
            # inspect, and log at ERROR so an inattentive operator can
            # find the failure.
            self.last_error = exc
            logger.error(
                "relay.flush.raise: dropped envelope (will be surfaced); "
                "error_type=%s error=%s",
                type(exc).__name__,
                str(exc),
            )
        else:
            self.processed_count += 1


__all__ = [
    "AsyncFlushDispatcher",
    "FlushMode",
    "FlushPolicy",
    "OnErrorMode",
]
